src/shogun/machine/GeneralizedLinearMachine.cpp

```diff
@@ -0,0 +1,299 @@
+/*
+ * STILL A WORK IN PROGRESS
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Tej Sukhatme
+ */
+
+// #include <rxcpp/rx-lite.hpp>
+// #include <shogun/features/DotFeatures.h>
+// #include <shogun/labels/Labels.h>
+// #include <shogun/labels/RegressionLabels.h>
+#include <shogun/machine/GeneralizedLinearMachine.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+// #include <utility>
+#include <math.h>
+#include <random>
+
+using namespace shogun;
+
+GeneralizedLinearMachine::GeneralizedLinearMachine(): Machine()
+{
+	init();
+}
+
+GeneralizedLinearMachine::GeneralizedLinearMachine(const std::shared_ptr<GeneralizedLinearMachine>& machine) : Machine()
+{
+	init();
+	require(machine, "No machine provided.");
+
+	auto w = machine->get_w();
+	auto w_clone = w.clone();
+	set_w(w_clone);
+	set_bias(machine->get_bias());
+}
+
+void GeneralizedLinearMachine::init()
+{
+	m_bias = 0;
+	features = NULL;
+	m_w = NULL;
+
+	distribution= POISSON;
+	m_alpha=0.5;
+	m_tau=NULL;
+	m_lambda=0.1;
+	m_learning_rate=2e-1;
+	m_max_iter=1000;
+	m_tolerance=1e-6;
+	m_eta=2.0;
+	m_fit_intercept=true;
+}
+
+
+GeneralizedLinearMachine::~GeneralizedLinearMachine()
+{
+
+}
+
+SGVector<float64_t> GeneralizedLinearMachine::predict(SGMatrix<float64_t> X)
+{
+	SGVector<float64_t> result = conditional_intensity(X, m_w, m_bias);
+	return result;
+}
+
+bool GeneralizedLinearMachine::fit(SGMatrix<float64_t> X, SGVector<float64_t> y)
+{
+	//Initialize parameters
+	unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+  	std::default_random_engine generator (seed);
+	
+  	std::normal_distribution<double> distribution (0.0,1.0);
+	int32_t n_features = X.num_rows;
+	SGVector<float64_t> w;
+	w.zero();
+	int bias = m_bias;
+	if (m_fit_intercept && m_bias == NULL)
+		bias = 1 / (n_features + 1) * distribution(generator);
+
+	if(m_w)
+		w = SGVector<float64_t>(m_w);
+	else
+	{
+		w = SGVector<float64_t>(n_features);
+		for (int i = 0; i < n_features; i++)
+			w = 1 / (n_features + 1) * distribution(generator);
+	}
+	
+	//Iterative Updates
+	int n_iter = 0;
+	for (int i = 0; i < m_max_iter; i++)
+	{
+		n_iter++;
+		SGVector<float64_t> w_old(m_w);
+		SGVector<float64_t> gradient_w = compute_grad_L2_loss_w(X, y, w, bias);
+		float64_t gradient_bias = compute_grad_L2_loss_bias(X, y, w, bias);
+
+		//Update
+        w = linalg::add(w, gradient_w, 1.0, -1*m_learning_rate);
+		if(m_fit_intercept)
+			bias = bias - m_learning_rate * gradient_bias;
+	
+		//Apply proximal operator
+		w = apply_proximal_operator(w, m_lambda * m_alpha);
```

Comment:
i wonder whether we can move this into another class? to separate inference from the model as done in other places in shogun? See the optimization classes

Reply:
So, you're suggesting I should make separate sub-classes of the GeneralizedLinearMachine class like Poisson Regression, etc which will have methods override these methods?

And move all this calculation related code to those classes?

Reply:
Or we can do something similar to how the implementation of the KNN classes is done?

Basically to create a pure virtual `GLMdistribution` class derives from LinearMachine and then to create `PoissonGLMdistribution`, `BinomialGLMdistribution` `GammaGLMdistribution`, etc classes as sub-classes of the `GLMdistribution` class.
These will just have the implementations of the calculation related functions like `non_linearity()`, `gradient_non_linearity()`, `compute_grad_L2_loss()` etc.

Then to call these, we will create an instance `distribution` of type `GLMdistribution` and call these methods from there depending on the distribution type(`distribution->non_linearity()`).

Reply:
no need to create new visible classes for every distribution.
We can do that with enums, it would just be nice to have all the code related to one distribution in one place, and then the enums simply call that.
But what I meant above is that you don't implement gradient descent yourself but call an existing solver instead

Reply:
> But what I meant above is that you don't implement gradient descent yourself but call an existing solver instead

Oh okay, Can you please give an example of another method where this has been done? I'm slightly confused.

Reply:
there is a lot of stuff in `optimization`. Check how that is used.